smartmeter_datacollector/smartmeter/siemens_td3511.py

```diff
@@ -0,0 +1,257 @@
+@dataclass
+class RegisterDataPoint:
+    obis: str
+    data_point_type: MeterDataPointType
+    scaling: float = 1.0
+
+
+DEFAULT_REGISTER_MAPPING = [
+    RegisterDataPoint("1.7.0", MeterDataPointTypes.ACTIVE_POWER_P.value, 1000),
+    RegisterDataPoint("2.7.0", MeterDataPointTypes.ACTIVE_POWER_N.value, 1000),
+    RegisterDataPoint("3.7.0", MeterDataPointTypes.REACTIVE_POWER_P.value, 1000),
+    RegisterDataPoint("4.7.0", MeterDataPointTypes.REACTIVE_POWER_N.value, 1000),
+    RegisterDataPoint("14.7", MeterDataPointTypes.NET_FREQUENCY.value),
+
+    RegisterDataPoint("31.7", MeterDataPointTypes.CURRENT_L1.value),
+    RegisterDataPoint("32.7", MeterDataPointTypes.VOLTAGE_L1.value),
+    RegisterDataPoint("81.7.4", MeterDataPointTypes.ANGLE_UI_L1.value, 3.141592653589793 / 180),
+
+    RegisterDataPoint("51.7", MeterDataPointTypes.CURRENT_L2.value),
+    RegisterDataPoint("52.7", MeterDataPointTypes.VOLTAGE_L2.value),
+    RegisterDataPoint("81.7.15", MeterDataPointTypes.ANGLE_UI_L2.value, 3.141592653589793 / 180),
+
+    RegisterDataPoint("71.7", MeterDataPointTypes.CURRENT_L3.value),
+    RegisterDataPoint("72.7", MeterDataPointTypes.VOLTAGE_L3.value),
+    RegisterDataPoint("81.7.26", MeterDataPointTypes.ANGLE_UI_L3.value, 3.141592653589793 / 180),
+
+    RegisterDataPoint("1.8.0", MeterDataPointTypes.ACTIVE_ENERGY_P.value, 1000),
+    RegisterDataPoint("1.8.1", MeterDataPointTypes.ACTIVE_ENERGY_P_T1.value, 1000),
+    RegisterDataPoint("1.8.2", MeterDataPointTypes.ACTIVE_ENERGY_P_T2.value, 1000),
+    RegisterDataPoint("2.8.0", MeterDataPointTypes.ACTIVE_ENERGY_N.value, 1000),
+    RegisterDataPoint("2.8.1", MeterDataPointTypes.ACTIVE_ENERGY_N_T1.value, 1000),
+    RegisterDataPoint("2.8.2", MeterDataPointTypes.ACTIVE_ENERGY_N_T2.value, 1000),
+    RegisterDataPoint("3.8.1", MeterDataPointTypes.REACTIVE_ENERGY_P.value, 1000),
+    RegisterDataPoint("4.8.1", MeterDataPointTypes.REACTIVE_ENERGY_N.value, 1000),
+]
+
+
+class SiemensParser():
+    REGEX = r"(.{3,20})\(([\d\-\.:]{3,20})[*\)](.{0,10}[^\)\r\n])?"
+
+    def __init__(self, use_system_time: bool = False) -> None:
+        self._use_system_time = use_system_time
+        self._meter_time = None
+        self._meter_date = None
+        self._timestamp = None
+        self._meter_id = None
+        self._buffer = []
+        self._register_obis = {r.obis: r for r in DEFAULT_REGISTER_MAPPING}
+
+    def append_to_buffer(self, received_data):
+        self._buffer.append(received_data.decode())
+
+    def clear_buffer(self):
+        self._buffer = []
+
+    def parse_data_objects(self, timestamp: datetime):
+        # Extract timestamp and meter id
+        self._timestamp = timestamp
+        for data in self._buffer:
+            result = re.search(SiemensParser.REGEX, data)
+            if result is None:
+                continue
+            obis, value, _ = result.groups()
+
+            # Extract meter id (common source id for all data points)
+            if obis == "0.0.0":
```

Comment:
Are the OBIS codes composed of only 3 digits instead of 6, here?

Reply:
Even worse, metrics are identified with obis codes varying between 2 and 4 numbers or single letters. That's why I extract the obis code as a whole for identification of the metric.